### ISSUE TITLE
Reference page content loads slowly

### ISSUE THREAD

**Issue by dhowe** [processing/p5.js-website#143]
When loading the reference page (http://p5js.org/reference/) there is often a noticeable lag where the  content is blank (~7-8 seconds for me), which may be confusing to new users and/or annoying...

![screen shot 2015-07-04 at 3 49 14 pm](https://cloud.githubusercontent.com/assets/737638/8507089/f2eab332-2265-11e5-8099-2b48f05c38c4.png)

(Note: I haven't tested this extensively, but it does appear to happen regardless of browser when the page is not cached)

OS X (10.10), FF (38), Chrome (43)

**Comment by dhowe** [comment]
Does no one else notice this? Several people in my part of the world (asia) have mentioned that the reference site was 'broken', due to this issue...

**Comment by spasmsaps** [comment]
@dhowe Yes, i also see this.
data.json is 843kb and is not minified.

![screenshot from 2015-07-21 15 04 53](https://cloud.githubusercontent.com/assets/4454698/8801361/5253328e-2fba-11e5-82a2-c4c3c8385410.png)

**Comment by lmccart** [comment]
yes this is an issue. I'll start by minifying data.json and see how much that helps. beyond that, anyone have any ideas?

**Comment by sepans** [comment]
Lazy loading?

(Just loading the top level function names for initialization, and making api calls for loading description and example when user clicked on a function)

**Comment by antiboredom** [comment]
I know it would be a big change but perhaps we could shift to using a static site generator for the reference.

**Comment by sepans** [comment]
lazy loading?

**Comment by lmccart** [comment]
@futuremarc not sure if you're working on any reference stuff but this would be a great one to look into. cc @shiffman

**Comment by futuremarc** [comment]
@lmccart I've been doing some messing around and found a way to download the entire reference for offline use, not sure how to turn it into a script to always package the latest reference, though.

[Here](https://www.dropbox.com/s/77c8s0kok70y0gv/Archive.zip?dl=0) is a zip of the current reference

**Comment by lmccart** [comment]
ah cool, this is great!! seems like we should be able to automate this somehow, we can try to take a look together sometime if that's helpful.

**Comment by dhowe** [comment]
Any progress on this? We are now starting to teach with p5.js and visiting this page and seeing nothing for 10 seconds is *very* confusing for new students

**Comment by futuremarc** [comment]
Ah, I'm going to to look into this

**Comment by lmccart** [comment]
thanks, that would be great! @futuremarc has also been working on a downloadable reference which should help at least a little.

**Comment by toolness** [comment]
@futuremarc, are you still looking into this?  I'm happy to help if you need it, no worries if not.

I'd like to second @antiboredom's suggestion of making the reference a statically-generated site: if it's all static assets, it could be deployed via CloudFront, which (I think?) would make it as fast for asia as it already is for the US/Europe. It would also make it a lot easier to package as a downloadable like futuremarc is suggesting.

Is PHP just being used as a templating engine? If so, there might be ways to generate a static site without needing to move away from PHP... Maybe this is what futuremarc's approach involves.

Barring a fully static site, though, another solution might be to just move all the non-dynamic (i.e., non-PHP) content onto CloudFront. That might be a bit annoying to migrate to, but at least it'd mean that assets like data.json would be downloaded as quickly from asia as they are from the US.

**Comment by futuremarc** [comment]
I actually haven't been looking into this. I do need to set aside an afternoon and put together the downloadable reference PR. Will probably focus on that during the next p5 developers meeting

**Comment by lmccart** [comment]
PHP is currently being used but we are planning to migrate away from PHP in the coming year to support internationalization. progress on that is over here:
https://github.com/mayaman/p5js-website

**Comment by limzykenneth** [comment]
@lmccart any progress on this? I did some test and from the original 930KB data.json, simply minifying gets it down to 49 KB, if it is GZIP the size goes all the way down to ~10KB. I don't really use grunt a lot so I can't really figure out where to put the extra step of minifying though...GZIPPing can wait until the node version of the site as it should be something the server do before serving the file.

Reducing the file size should make rendering the reference fast enough that we don't really need to do anything else too tricky I think.

**Comment by limzykenneth** [comment]
I did a bit more tests and there are two bottle necks both file loads, namely reference.js and data.json. The problem is that reference.js is quite big itself and data.json will only load when reference.js have loaded and executed the ajax call for data.json. That results in a user having to load ~1.18mb of files before the reference appear. Minifying them certainly improves load time but not by very much.

One way to improve the load times will be to get underscore and backbone out of reference.js and potentially load them via cdn. That way they can either be loaded in parallel (might have some problems) or loaded in the header, reducing the overall file size of reference.js.

**Comment by JobLeonard** [comment]
I think one of the most amazing and underrated features of the original Processing has always been that it comes with all of it's documentation in an offline, static website, which was then easy to access from the editor. I would love for p5 to continue this tradition of being super-easy to play with and explore at any time, regardless of internet access.

So what I'm saying is: please don't underestimate the value of an offline pre-generated version of the docs.

**Comment by dhowe** [comment]
I agree with @JobLeonard . Especially for those of us in Asia, where it can take 10 seconds to access the docs, the lack of a pre-generated ref is just about a deal-breaker for teaching...

**Comment by limzykenneth** [comment]
I might have a go at this in a couple weeks time when I'm done with my own stuff. Plan is to improve load speed on the web and to have a static version generated on build so that it can be accessed offline.

**Comment by lmccart** [comment]
Yes, I agree this is very important. Things are a bit hectic for me right now, but I will take a stab at it as soon as I can if @limzykenneth doesn't get to it first. Thanks for your patience all!
